File: algorithms/python/26_remove_duplicates_from_sorted_array/remove_duplicates_from_sorted_array.py
"""
26. remove duplicates from sorted array

https://leetcode-cn.com/problems/remove-duplicates-from-sorted-array

删除排序数组中的重复项
"""

from typing import List


def remove_duplicates(nums: List[int]) -> int:
    """
    dual pointer(快慢指针)

    Complexity Analysis
    time complexity: O(N)
    space complexity: O(1)
    """

    if not nums:
        return 0

    n = len(nums)
    fast = slow = 1
    while fast < n:
        if nums[fast] != nums[fast - 1]:
            nums[slow] = nums[fast]
            slow += 1
        fast += 1

    return slow

File: algorithms/go/26-remove-duplicates-from-sorted-array/remove-duplicates-from-sorted-array.go
package package26

// removeDuplicates dual pointer(快慢指针)
func removeDuplicates(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	n := len(nums)
	fast, slow := 1, 1
	for fast < n {
		if nums[fast] != nums[fast-1] {
			nums[slow] = nums[fast]
			slow++
		}
		fast++
	}
	return slow
}
